### DISCUSSION TITLE
Azure namespace migration from Microsoft.Web to Microsoft.App in March 2022

### DISCUSSION THREAD

**Discussion by danielrbradley** [pulumi/pulumi#9099] [Pulumi Providers] [answered]
As announced by Microsoft (https://github.com/microsoft/azure-container-apps/issues/109), all Container App resources will be migrated from the `Microsoft.Web` namespace to the `Microsoft.App` namespace during March 2022.

The new API namespace is already available in each Pulumi SDK:

- [Dotnet](https://github.com/pulumi/pulumi-azure-native/tree/master/sdk/dotnet/App)
- [Go](https://github.com/pulumi/pulumi-azure-native/tree/master/sdk/go/azure/app)
- [NodeJS](https://github.com/pulumi/pulumi-azure-native/tree/master/sdk/nodejs/app)
- [Python](https://github.com/pulumi/pulumi-azure-native/tree/master/sdk/python/pulumi_azure_native/app)

# Upgrade Instructions

Pulumi sees this migration as the old resources being removed and the new resources being created. Therefore to replicate the migration in our Pulumi state we're doing two steps:

1. Drop the resources being migrated from being managed by Pulumi.
2. Importing the new migrated resources back into our Pulumi stack

## Preparation

Add the [`retainOnDelete` resource option](https://www.pulumi.com/docs/intro/concepts/resources/options/retainondelete/) to the resources which are going to be migrated.

This option allows us to remove the code for the resources once they've changed namespace without attempting to delete the underlying cloud resources. The resources will only be removed from our Pulumi stack state.

## Migration

Once the cloud resources have been migrated by Microsoft, update your resources (examples below in Typescript but should be approximately the same in any other supported Pulumi language).

1. Update your imports:

   ```diff
   - import * as web from "@pulumi/azure-native/web";
   + import * as app from "@pulumi/azure-native/app";
   ```

2. Update resource namespaces based on resource update:

   - Update the import namespace.
   - Change `KubeEnvironment` to `ManagedEnvironment`
   - **Important**: change the Pulumi ID for the resource.

   ```diff
   - const environment = new web.KubeEnvironment("kubeEnvironment", {
   + const environment = new app.ManagedEnvironment("managedEnvironment", {
   ```

   ```diff
   - const containerApp = new web.ContainerApp("containerApp", {
   + const containerApp = new app.ContainerApp("containerAppImported", {
   ```

3. Fix arguments which have been renamed e.g.

   ```diff
   const containerApp = new app.ContainerApp("containerAppImported", {
   -   kubeEnvironmentId: environment.id,
   +   managedEnvironmentId: environment.id,
   ```

4. Add import options for the migrated resources

   - Add the `import` option with the Resource ID for each resource (this can be found in the JSON view in the portal).

   ```diff
   const kubeEnvironment = new app.ManagedEnvironment(
     "managedEnvironment",
     {
       // args
     },
   +   {
   +     // opts
   +     import:
   +       "/subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/resourceGroupXXXXXXXX/providers/Microsoft.App/managedenvironment+ s/environmentcXXXXXXX",
   +     ignoreChanges: [
   +       "appLogsConfiguration.logAnalyticsConfiguration.sharedKey", // this is a write-only property
   +     ],
   +   }
   );
   const containerApp = new app.ContainerApp(
     "containerAppImported",
     {
       // args
     },
   +   {
   +     // opts
   +     import:
   +       "/subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/resourceGroupXXXXXXXX/providers/Microsoft.App/containerApps/container-app",
   +     ignoreChanges: [ // Default values
   +       "configuration.activeRevisionsMode",
   +       "configuration.ingress.allowInsecure",
   +       "template.scale",
   +     ],
   +   }
   );
   ```

   Tip: if you get a message saying "inputs to import do not match the existing resource" during preview, expand the details to show a diff of what doesn't match. You can fix these import differences by changing your code to match, or adding to the `ignoreChanges` options if it's just a default value.

5. Run a pulumi deployment to complete the import.
6. Remove the `import` and `ignoreChanges` before the next deploy.
7. If you'd like the resources to have their original name, [add an alias](https://www.pulumi.com/docs/intro/concepts/resources/options/aliases/) for the old name (this can be removed again once a deploy has been completed):

   ```diff
   const containerApp = new app.ContainerApp(
   -   "containerAppImported",
   +   "containerApp",
     {
       // args
     },
   +   {
   +     // opts
   +     alias: [{ name: "containerAppImported" }]
   +   }
   );
   ```

Please post any questions you may have regarding the migration in this discussion thread.

**Comment by michaelmcneilnet** [comment, marked as answer]
C# Pulumi.AzureNative v1.61.0

While trying this migration, I receive the following error:

```
Diagnostics:
  azure-native:app:ManagedEnvironment (ManagedEnvironment):
    error: Preview failed: autorest/azure: Service returned an error. Status=400 Code="NoRegisteredProviderFound" Message="No registered resource provider found for location 'northeurope' and API version '2022-01-01-preview' for type 'kubeEnvironments'. The supported api-versions are '2021-03-01, 2021-02-01, 2021-01-15, 2021-01-01, 2020-12-01, 2020-10-01, 2020-09-01, 2020-06-01, 2019-08-01, 2019-02-01, 2019-01-01, 2018-11-01, 2018-08-01, 2018-05-01-preview, 2018-02-01, 2017-08-01, 2016-09-01, 2016-03-01, 2015-08-01, 2015-07-01, 2015-06-01, 2015-05-01, 2015-04-01, 2015-02-01, 2014-11-01, 2014-06-01, 2014-04-01'. The supported locations are 'northcentralusstage, westcentralus, eastus, westeurope, jioindiawest, northeurope, canadacentral, eastus2'."
```
I was surprised to see type `kubeEnvironments` being used and not `managedEnvironments`.

If I use the Azure Portal to create a new ACA via the UI, I can see in the template that it is pointing to the new namespace, so I assume it's available in northeurope?

```json
{
    "$schema": "http://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#",
    "contentVersion": "1.0.0.0",
    "parameters": {
        "subscriptionId": {
            "type": "string"
        },
        "name": {
            "type": "string"
        },
        "location": {
            "type": "string"
        },
        "environmentId": {
            "type": "string"
        },
        "containers": {
            "type": "array"
        },
        "secrets": {
            "type": "array"
        },
        "registries": {
            "type": "array"
        },
        "ingress": {
            "type": "object"
        },
        "environmentName": {
            "type": "string"
        },
        "workspaceName": {
            "type": "string"
        },
        "workspaceLocation": {
            "type": "string"
        }
    },
    "resources": [
        {
            "name": "[parameters('name')]",
            "kind": "containerapps",
            "location": "[parameters('location')]",
            "dependsOn": [
                "[concat('Microsoft.App/managedEnvironments/', parameters('environmentName'))]"
            ],
            "properties": {
                "configuration": {
                    "secrets": "[parameters('secrets')]",
                    "registries": "[parameters('registries')]",
                    "ingress": "[parameters('ingress')]"
                },
                "template": {
                    "containers": "[parameters('containers')]"
                },
                "managedEnvironmentId": "[parameters('environmentId')]"
            },
            "apiVersion": "2022-01-01-preview",
            "type": "Microsoft.App/containerapps"
        },
        {
            "name": "[parameters('environmentName')]",
            "location": "[parameters('location')]",
            "dependsOn": [
                "[concat('Microsoft.OperationalInsights/workspaces/', parameters('workspaceName'))]"
            ],
            "properties": {
                "internalLoadBalancerEnabled": false,
                "appLogsConfiguration": {
                    "destination": "log-analytics",
                    "logAnalyticsConfiguration": {
                        "customerId": "[reference(concat('Microsoft.OperationalInsights/workspaces/', parameters('workspaceName')), '2020-08-01').customerId]",
                        "sharedKey": "[listKeys(concat('Microsoft.OperationalInsights/workspaces/', parameters('workspaceName')), '2020-08-01').primarySharedKey]"
                    }
                }
            },
            "apiVersion": "2022-01-01-preview",
            "type": "Microsoft.App/managedEnvironments"
        },
        {
            "apiVersion": "2020-08-01",
            "name": "[parameters('workspaceName')]",
            "type": "Microsoft.OperationalInsights/workspaces",
            "location": "[parameters('workspaceLocation')]",
            "dependsOn": [],
            "properties": {
                "sku": {
                    "name": "PerGB2018"
                },
                "retentionInDays": 30,
                "workspaceCapping": {}
            }
        }
    ]
}
```

Can I also check that the import string I should be using is the current Microsoft.Web namespace that my existing ACA has, taken from the JSON view in the portal? I noticed in the guide above that it is importing from the Microsoft.App namespace.

For example, for the managed environment I am using 

```cs
new CustomResourceOptions
{
    ImportId = $"/subscriptions/XXX/resourceGroups/XXX/providers/Microsoft.Web/kubeEnvironments/XXX",
    IgnoreChanges = new List<string>
    {
        "appLogsConfiguration.logAnalyticsConfiguration.sharedKey"
    },
    RetainOnDelete = true
});
```

**Reply by danielrbradley** [reply]
If you're importing from the Microsoft.Web namespace, you'll still need to import the resources using the classes from the `Pulumi.AzureNative.Web` namespace. After your resources have been moved by Microsoft to the `Microsoft.App` namespace, then you'll be able to import the resources using the `Pulumi.AzureNative.App` namespace.

The types used in your program have to correspond to the namespace where the Azure resources currently belong (which will then change when they are migrated).

If you're creating new resources from scratch, you can start using the `App` namespace straight away.

Hope that helps clarify the process.

**Reply by michaelmcneilnet** [reply]
Thanks for the help, getting there. 

The last issue seems to be that the `configuration.secrets` section keeps reporting `inputs to import do not match the existing resource` even after making sure they are the same as before. Can I ignore this section in `ignoreChanges` or will this cause issues?

Update - It seems that because the json representing the ACA resource only specifies the name of the secret, but my Pulumi code of course has the value as well as the name, the import thinks there is a mismatch. Removing the secret values from my pulumi program and leaving only the names seems to stop the mismatch warning.

**Reply by danielrbradley** [reply]
Yes I expect an ignore would be best to use there.
